packages/ra-auth-cognito/src/authProvider.ts

```diff
@@ -77,37 +93,83 @@ export const CognitoAuthProvider = (
+                        reject(new ErrorRequireNewPassword());
+                    },
+                    mfaSetup: () => {
+                        user.associateSoftwareToken({
+                            associateSecretCode: secretCode => {
+                                reject(
+                                    new ErrorMfaTotpAssociationRequired({
+                                        secretCode,
+                                        username: user.getUsername(),
+                                        applicationName:
+                                            config.applicationName ??
+                                            window.location.hostname,
+                                    })
+                                );
+                            },
+                            onFailure: err => {
+                                reject(err);
+                            },
+                        });
+                    },
+                    totpRequired: () => {
+                        reject(new ErrorMfaTotpRequired());
+                    },
+                    mfaRequired: () => {
+                        reject(new ErrorMFARequired());
+                    },
+                };
+                if (formIsNewPassword(form)) {
+                    const { newPassword, confirmNewPassword, ...attributes } =
+                        form;
                     if (!user) {
                         return reject(new Error('User is not defined'));
                     }
 
                     return user.completeNewPasswordChallenge(
                         newPassword,
                         attributes,
-                        {
-                            onSuccess: result => {
-                                resolve(result);
-                            },
-                            onFailure: err => {
-                                reject(err);
-                            },
-                        }
+                        callback
+                    );
+                }
+                if (formIsTotp(form)) {
+                    if (!user) {
+                        return reject(new Error('User is not defined'));
+                    }
+                    return user.sendMFACode(
+                        form.totp,
+                        callback,
+                        'SOFTWARE_TOKEN_MFA'
+                    );
+                }
+
+                if (formIsTotpAssociation(form)) {
+                    if (!user) {
+                        return reject(new Error('User is not defined'));
+                    }
+                    return user.verifySoftwareToken(
+                        form.association,
+                        'Authenticator',
+                        callback
                     );
                 }
 
+                if (!formIsLogin(form)) {
+                    return reject(new Error('Invalid form'));
+                }
```

Comment:
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question.  -->
**issue:** ‏Unless I'm mistaken, this is unreachable code, which can therefore be removed

Reply:
I changed the code in `formIsLogin` to verify he form.